index.d.ts

```diff
@@ -137,14 +42,844 @@ declare class Enquirer<T = object> extends EventEmitter {
+      } & types.QuestionBase &
+        types.Initializer<string | number, string[]> &
+        types.Formatter<string, string[]>
+    }
+
+    export type FormQuestion = {
+      type: 'form',
+      choices: FormQuestion.Choice[],
+      align?: 'left' | 'right',
+    } & types.QuestionBase &
+      types.Validator<FormQuestion.Answer, FormQuestion.Answer> &
+      types.ResultTransformer<FormQuestion.Answer, FormQuestion.Answer>
+
+    export namespace FormQuestion {
+      export type Choice = ChoiceOptions | Promise<ChoiceOptions> | (() => ChoiceOptions | Promise<ChoiceOptions>)
+      export type ChoiceOptions = {
+        name: string,
+        value?: string,
+        message: string,
+        hint?: string,
+        initial?: string,
+        disabled?: boolean
+      }
+      export type Answer = Record<string, string>
+    }
+  }
+  export namespace types {
+    export type Value = string | boolean | number | string[] |
+      prompt.ScaleQuestion.Answer | prompt.SnippetQuestion.Answer | prompt.FormQuestion.Answer
+
+    export type Answer = string | boolean | number | string[] |
+      prompt.QuizQuestion.Answer | prompt.ScaleQuestion.Answer | prompt.SnippetQuestion.Answer |
+      prompt.FormQuestion.Answer
+
+    export type QuestionBase = {
+      name: string,
+      message: string | (() => string | Promise<string>),
+
+      skip?: boolean | ((state: any) => boolean | Promise<boolean>),
+      show?: boolean,
+      onSubmit?: (name: string, value: string, prompt: Prompt) => boolean | Promise<boolean>,
+      onCancel?: (name: string, value: string, prompt: Prompt) => boolean | Promise<boolean>,
+    }
+
+    export type Initializer<V extends types.Value, A extends types.Answer> = {
+      initial?: V | ((this: Prompt<A>) => V | Promise<V>),
+    }
+
+    export type Formatter<V extends types.Value, A extends types.Answer> = {
+      format?: (this: Prompt<A>, value: V | undefined) => string | Promise<string>,
+    }
+
+    export type Validator<V extends types.Value, A extends types.Answer> = {
+      validate?: (this: Prompt<A>, value: V) => boolean | string | Promise<boolean | string>,
+    }
+
+    export type ResultTransformer<V extends types.Value, A extends types.Answer> = {
+      result?: (this: Prompt<A>, value: V) => A | Promise<A>,
+    }
+
+    export type Key = {
+      name?: string,
+      code?: string,
+      sequence?: string,
+      ctrl?: boolean,
+      shift?: boolean,
+      fn?: boolean,
+    }
+
+    export type ChoiceInput = string | Promise<string> | ChoiceOptions | (() => string | Promise<string>)
+
+    export type ChoiceOptions = {
+      name?: string,
+      message?: string,
+      hint?: string,
+      disabled?: boolean,
+      value?: types.Answer,
+    }
+
+    export type Choice = {
+      name: string,
+      message: string,
+      hint?: string,
+      disabled?: boolean,
+      value?: types.Answer,
```

Comment:
Shouldn't this type be open-ended so the user can supply custom properties to a `Choice`?

```suggestion
      value?: types.Answer,
      [key:string]: unknown
```

Reply:
Not sure about this one. I didn't see how this is used by the user.
Does the choice object pass to one of the callbacks (initialize, format, validate, result)?